### DISCUSSION TITLE
Contact listener determinism after deserialization

### DISCUSSION THREAD

**Discussion by opera-aberglund** [louis-langholtz/PlayRho#572] [General]
I encountered some issues with desynchronized state after serialization/deserialization of the world when using contact listeners and creating and destroying some joints based on these callbacks. The issue has been hard to pinpoint, but I have some general questions about contact listeners either way.

For example, if two bodies are currently "in contact" and the BeginContact event has already been triggered, what would happen after serialization/deserialization of the state; would the "BeginContact" event be triggered again? Where is this information about what events have already been triggered stored?

**Comment by louis-langholtz** [comment]
Hi @opera-aberglund , thank you for asking!

The questions you've stated are:

1. If two bodies are currently "in contact" and the BeginContact event has already been triggered, what would happen after serialization/deserialization of the state; would the "BeginContact" event be triggered again?
2. Where is this information about what events have already been triggered stored?

Answers:

1. I believe that after serialization/deserialization the "BeginContact" event would be triggered _again_ in the destination world. I will check this to confirm. Note that if doing the special zero time step at the end of deserialization, it seems like one should be able to recognize that special state such that either listeners aren't called, or the listeners effectively ignore changes during that step. The net effect of which I believe would be that an event like "BeginContact" doesn't need to trigger the same effect again on the destination world.
2. State for contact beginning and contact ending is effectively stored in the associated manifolds. I.e. the manifold at the `ContactID` of the contact.

**Reply by opera-aberglund** [reply]
Thanks for your answer! 

Yes, that is also the behaviour I'm suspecting, and it also seems like it should be fairly easy to work around it. For now, I'm just ignoring the listeners for my game since I don't really need them at the moment, but I might investigate this more later if it can cause desynchronisation.

**Reply by louis-langholtz** [reply]
Incidentally, I've also setup `[playrho]` as a tag on StackOverflow. I'd love to get more [questions over on StackOverflow that are tagged with that](https://stackoverflow.com/questions/tagged/playrho)!

Admittedly, I don't check it regularly but if you are open to putting any questions there, you're welcome to also let me know on GitHub via a link or something in an issue or discussion.
